spec/outputs/loggly_spec.rb

```diff
@@ -24,42 +34,165 @@
+      it 'should perform one http request per batch of common key+tag' do
+        config['key'] = '%{custom_key}'
+        config['tag'] = '%{custom_tag}'
+        event.set('custom_key', 'generally_used_key')
+
+        event1 = event.clone.tap { |e| e.set('message', 'event1') }
+        event2 = event.clone.tap { |e| e.set('message', 'event2') ; e.set('custom_key', 'other_key') }
+        event3 = event.clone.tap { |e| e.set('message', 'event3') ; e.set('custom_tag', 'other_tag') }
+        event4 = event.clone.tap { |e| e.set('message', 'event4') }
+
+        expect(output).to receive(:perform_api_call) { |url, body|
+          expect(body).to match /"event1"/
+          expect(body).to match /"event4"/
+          expect(url).to eq('http://logs-01.loggly.com/bulk/generally_used_key/tag/logstash')
+        }
+        expect(output).to receive(:perform_api_call) { |url, body|
+          expect(body).to match /"event2"/
+          expect(url).to eq('http://logs-01.loggly.com/bulk/other_key/tag/logstash')
+        }
+        expect(output).to receive(:perform_api_call) { |url, body|
+          expect(body).to match /"event3"/
+          expect(url).to eq('http://logs-01.loggly.com/bulk/generally_used_key/tag/other_tag')
+        }
+        expect(output).not_to receive(:perform_api_call) # anymore
+
+        output.multi_receive([event1, event2, event3, event4])
+      end
+    end
+  end
+
+  context 'splitting batches of events' do
+    context 'when they are all with the same key+tag' do
+      it 'should return one batch' do
+        batches = output.split_batches([ {event: :event1, key: 'key1', tag: 'tag1'},
+                                         {event: :event2, key: 'key1', tag: 'tag1'} ])
+        expect(batches.size).to eq(1)
+        expect(batches).to eq({ ['key1', 'tag1'] => [:event1, :event2] })
+      end
+    end
+
+    context 'when messages have different key & tag' do
+      it 'should return one batch for each key+tag combination' do
+        batches = output.split_batches([
+          {event: :event1, key: 'key1', tag: 'tag1'},
+          {event: :event2, key: 'key2', tag: 'tag1'},
+          {event: :event3, key: 'key2', tag: 'tag2'},
+          {event: :event4, key: 'key1', tag: 'tag1'},
+          {event: :event5, key: 'key2', tag: 'tag1'},
+          {event: :event6, key: 'key1', tag: 'tag1'},
+          {event: :event7, key: 'key1', tag: 'tag1'},
+        ])
+        expect(batches.size).to eq(3)
+        expect(batches).to eq(
+          { ['key1', 'tag1'] => [:event1, :event4, :event6, :event7],
+            ['key2', 'tag1'] => [:event2, :event5],
+            ['key2', 'tag2'] => [:event3],
+          })
+      end
+    end
+  end
+
+  context 'when building message bodies' do
+    it 'should send only one payload when everything fits' do
+      yielded_times = 0
+      output.build_message_bodies([event] * 10) do |body|
+        insist { body.lines.count } == 10
```

Comment:
we're moving more and more away from insist, so for new tests I suggest using instead `expect(body.lines.count).to eq(10)`

Reply:
Makes sense. Less magic to have the `eq()` matcher. I'll change that.